security/advancedtls/advancedtls_integration_test.go

```diff
@@ -408,3 +411,272 @@ func (s) TestEnd2End(t *testing.T) {
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer clientIdentityProvider.Close()
+			clientRootProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				TrustFile:    clientTrustTmp.Name(),
+				RootInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer clientRootProvider.Close()
+			serverIdentityProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				CertFile:         serverCertTmp.Name(),
+				KeyFile:          serverKeyTmp.Name(),
+				IdentityInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer serverIdentityProvider.Close()
+			serverRootProvider, err := NewPEMFileProvider(PEMFileProviderOptions{
+				TrustFile:    serverTrustTmp.Name(),
+				RootInterval: 200 * time.Millisecond,
+			})
+			if err != nil {
+				t.Errorf("NewPEMFileProvider failed: %v", err)
+			}
+			defer serverRootProvider.Close()
+			// Start a server and create a client using advancedtls API with Provider.
+			serverOptions := &ServerOptions{
+				IdentityOptions: IdentityCertificateOptions{
+					IdentityProvider: serverIdentityProvider,
+				},
+				RootOptions: RootCertificateOptions{
+					RootProvider: serverRootProvider,
+				},
+				RequireClientCert: true,
+				VerifyPeer: func(params *VerificationFuncParams) (*VerificationResults, error) {
+					return &VerificationResults{}, nil
+				},
+				VType: CertVerification,
+			}
+			serverTLSCreds, err := NewServerCreds(serverOptions)
+			if err != nil {
+				t.Fatalf("failed to create server creds: %v", err)
+			}
+			s := grpc.NewServer(grpc.Creds(serverTLSCreds))
+			defer s.Stop()
+			lis, err := net.Listen("tcp", port)
+			if err != nil {
+				t.Fatalf("failed to listen: %v", err)
+			}
+			defer lis.Close()
+			pb.RegisterGreeterServer(s, greeterServer{})
+			go s.Serve(lis)
+			clientOptions := &ClientOptions{
+				IdentityOptions: IdentityCertificateOptions{
+					IdentityProvider: clientIdentityProvider,
+				},
+				VerifyPeer: func(params *VerificationFuncParams) (*VerificationResults, error) {
+					return &VerificationResults{}, nil
+				},
+				RootOptions: RootCertificateOptions{
+					RootProvider: clientRootProvider,
+				},
+				VType: CertVerification,
+			}
+			clientTLSCreds, err := NewClientCreds(clientOptions)
+			if err != nil {
+				t.Fatalf("clientTLSCreds failed to create, error: %v", err)
+			}
+			// At initialization, the connection should be good.
+			ctx1, cancel1 := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel1()
+			conn, greetClient, err := callAndVerifyWithClientConn(ctx1, "rpc call 1", clientTLSCreds, false)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer conn.Close()
+			// Make the identity cert change, and wait 1 second for the provider to
+			// pick up the change.
+			test.certUpdateFunc()
+			time.Sleep(1 * time.Second)
```

Comment:
These sleeps can quickly add up. Is there a way to avoid this?

One possible way is to define a helper func in `pemfile_provider.go` which wraps the call to `Distributor.Set()`. And from the test you can override that function to call the original function, and also push on a channel that the test is blocked on. So, instead of sleeping, you would block on that channel. And we have a `testutils.Channel` type which provided a channel receive operation to which you can pass a context with a deadline.

Reply:
Yeah I think it's possible to use a channel to avoid `sleep`. The problem is, this `Distributor.Set()` is run periodically by a goroutine. If we wrap it to push a message to a channel, the channel will soon be blowed up.
I have tried using a bool value and then wait on the main logic for that bool value to be flipped, but it seems this boolean value needs to be carefully managed(we will set it up before we change the file contents, and then use a for loop to wait for that value to be flipped). That might make the code harder to understand. 
I've adjusted the `sleep` interval to a smaller value. Would you mind if we keep it as it is right now?